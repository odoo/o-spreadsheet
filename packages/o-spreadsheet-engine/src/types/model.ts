import { GeoChartRegion } from "./chart/geo_chart";
import { Client, ClientPosition } from "./collaborative/session";
import { TransportService } from "./collaborative/transport_service";
import { Currency } from "./currency";
import { InformationNotification } from "./env";
import { FileStore } from "./files";
import { Locale } from "./locale";
import { Color } from "./misc";

/**
 * Model
 *
 * The Model class is the owner of the state of the Spreadsheet. However, it
 * has more a coordination role: it defers the actual state manipulation work to
 * plugins.
 *
 * At creation, the Model instantiates all necessary plugins. They each have
 * a private state (for example, the Selection plugin has the current selection).
 *
 * State changes are then performed through commands.  Commands are dispatched
 * to the model, which will then relay them to each plugins (and the history
 * handler). Then, the model will trigger an 'update' event to notify whoever
 * is concerned that the command was applied (if it was not cancelled).
 *
 * Also, the model has an unconventional responsibility: it actually renders the
 * visible viewport on a canvas. This is because each plugins actually manage a
 * specific concern about the content of the spreadsheet, and it is more natural
 * if they are able to read data from their internal state to represent it on the
 * screen.
 *
 * Note that the Model can be used in a standalone way to manipulate
 * programmatically a spreadsheet.
 */
export type Mode = "normal" | "readonly" | "dashboard";

export interface ModelConfig {
  readonly mode: Mode;
  /**
   * Any external custom dependencies your custom plugins or functions might need.
   * They are available in plugins config and functions
   * evaluation context.
   */
  readonly custom: Readonly<{
    [key: string]: any;
  }>;
  readonly defaultCurrency?: Partial<Currency>;
  /**
   * External dependencies required to enable some features
   * such as uploading images.
   */
  readonly external: Readonly<ModelExternalConfig>;
  readonly moveClient: (position: ClientPosition) => void;
  readonly transportService: TransportService;
  readonly client: Client;
  readonly snapshotRequested: boolean;
  readonly notifyUI: (payload: InformationNotification) => void;
  readonly raiseBlockingErrorUI: (text: string) => void;
  readonly customColors: Color[];
}

export interface ModelExternalConfig {
  readonly fileStore?: FileStore;
  readonly loadCurrencies?: () => Promise<Currency[]>;
  readonly loadLocales?: () => Promise<Locale[]>;
  readonly geoJsonService?: {
    getAvailableRegions: () => GeoChartRegion[];
    getTopoJson: (region: string) => Promise<any>;
    /**  Convert the name of a geographical feature (eg. France) to the id of the corresponding feature in the TopoJSON */
    geoFeatureNameToId: (region: string, territory: string) => string | undefined;
  };
}
