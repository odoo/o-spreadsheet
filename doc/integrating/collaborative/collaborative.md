# Collaborative Edition

Realtime collaboration edition can be enabled to synchronize a spreadsheet across multiple clients.
It is enabled by providing a way to communicate with other connected clients to the `<Spreadsheet/>` component. We call it the _transport service_. Its interface and how it should be implemented is described in a [dedicated section](#transport-service). An optional [client](tsdoc/interfaces/client.md) can also be provided to display the names of connected clients.

```ts
const model = new Model(data, {
  transportService,
  client: { id: 456, name: "Raoul" },
});
```

```xml
<Spreadsheet model="model"/>
```

To learn more about the explored possibilities, please read [this](collaborative_choices.md).

- [Collaborative Edition](#collaborative-edition)
  - [Transport Service](#transport-service)
  - [Coordinating server](#coordinating-server)
  - [How it works](#how-it-works)
    - [Operational Transform](#operational-transform)
  - [Extension](#extension)
    - [Commands](#commands)
    - [Transformations](#transformations)
    - [History](#history)

## Transport Service

To enable realtime collaboration edition, a TransportService should be given to the Spreadsheet component.

The TransportService is responsible for sending and receiving messages between the clients with the help of central server.

The transport service should implement the [TransportService](tsdoc/interfaces/transportservice.md) interface.

An example can be found in demo/transport.js. The example is implemented using Websockets.

## Coordinating server

[Messages](tsdoc/README.md#CollaborationMessage) sent through the transport service should be handled by a centralized server. Its goal it to coordinate messages, order them and maintain the current spreadsheet revision.
For each received messages, the server decides if it's accepted or not. If the message is accepted, it's transmitted to all clients
(including the one which sent the message).

Messages can be split in two categories:

1. The messages that do not change the state of the spreadsheet and are always accepted:

- [ClientJoinedMessage](tsdoc/interfaces/clientjoinedmessage.md)
- [ClientMovedMessage](tsdoc/interfaces/clientmovedmessage.md)
- [ClientLeftMessage](tsdoc/interfaces/clientleftmessage.md)

1. The messages that require special care to ensure a correct message ordering:

- [RemoteRevisionMessage](tsdoc/interfaces/remoterevisionmessage.md)
- [RevisionUndoneMessage](tsdoc/interfaces/revisionundonemessage.md)
- [RevisionRedoneMessage](tsdoc/interfaces/revisionredonemessage.md)

They have two important properties: `serverRevisionId` and `nextRevisionId`. A message is accepted only if its `serverRevisionId`
is the same as the server revision id. The message is then transmitted to all clients, and the server revision id becomes `nextRevisionId`. Otherwise, the message is rejected. It means that the client should re-sent it later, after transforming it with all the messages it received in the meantime.

By convention, the initial revision id of a blank spreadsheet is `"START_REVISION"`.

A basic example can be found in tools/server/main.js. The example is implemented using Websockets

## How it works

Realtime collaborative editing is based on [Operational Transform](https://en.wikipedia.org/wiki/Operational_transformation).

### Operational Transform

The aim of operation transform is to manage realtime collaborative editing on a document and preserve consistency even with concurrent operations.
It ensures all clients eventually converge to the same state while preserving user intentions.

Let's take a look at two concurrent operations, why it can lead to problems and how operational transform can help:

Alice and Bob are working on the same sheet. At the same time, Alice wants to add a column before "B" and Bob wants to update the cell "B1" with the content "Hello". Let's call Alice's operation `OA` and Bob's operation `OB`.
Both Alice and Bob send their operation to the other.

When Alice receives `OB`, it would update "B1" with "Hello". When Bob receives `OA`,
it would add a column before "B", which moves the current column B to column C.
We now have:

- Alice with an additional column before "B" and "Hello" in "B1"
- Bob with an additional column before "B" and "Hello" in "C1"

Both state have diverged. This can't be good...

That's when operational transform can help.

`OB` on Alice's side is no longer valid since column B was moved to column C. The operation should be adapted to
reflect that. To do that, we apply a transformation to `OB`, to target the column C instead of column B. `OB` becomes "Update "C1" with the content `Hello`". With this transformation, Bob's intention is preserved and the states converge.

In order to identify and resolve which operations conflict, we need a way to detect two concurrent actions, and a way to order them.

To detect concurrency, we introduce a revision log, which is a unique identifier to indicate on which state the action is executed.
Each time an action is received and accepted by the server (the revision log of the server is the same as the action), the revision log is incremented.

Now, we need a way to transform the commands. An easy way would be to transform commands on the server, like [Google Drive](https://drive.googleblog.com/2010/09/whats-different-about-new-google-docs.html). But, in the context of using o-spreadsheet with Odoo, we didn't want to duplicate the transformations functions in both client and server, as the two are implemented in different language.

When a client executes an action, it's locally executed, sent to the server and kept in the pending actions of the user.
If the action is accepted by the server, the action is removed from the pending ones. In the other case, the pending actions are reverted, transformed with the action received from the server, re-applied locally and resent to the server until they are accepted.

Local undo is implemented by keeping locally the actions of the user, revert to just before the undo-ed action, transform the next actions as if the undo-ed action was not executed and re-apply them. Local redo follow the same logic.

To keep a synchronous state, we only need to share the commands which impacts the state of spreadsheet (columns, grid, ...) but not the local state (selection of the user, composer state, ...)

This solution has a lot of pros, but also some cons:

1. We need to write a transformation function for each command we create, which could theoretically becomes huge. However, in practice, the transformations only concerns commands which changes the grid (add/remove columns, remove a sheet, merge).
1. Undo/Redo is synchronous, i.e. it should be accepted by the server before being executed locally.

## Extension

To extend o-spreadsheet without breaking realtime collaborative edition, there are some points to keep in mind.

### Commands

There are two types of [commands](add_command.md): `CoreCommands` and `Commands`. Only `CoreCommands` are synchronized.
Here is the way to register a `CoreCommand` to o-spreadsheet.

```js
const { coreTypes } = o_spreadsheet;

coreTypes.add("MY_COMMAND_NAME");
```

### Transformations

For each new `CoreCommand`, a transformation function could be required for each other `CoreCommands`.

A transformation function takes as arguments the already executed command, and the command to transform. It should return a `CoreCommands` or `undefined` if the command should be skipped.

If a transformation is required, here is the way to declare it. The transformation check that the command is on the same sheet of the deleted sheet. If true, the command should be skipped.

```js
const { otRegistry } = o_spreadsheet.registries;

otRegistry.addTransformation("DELETE_SHEET", ["MY_COMMAND"], (myCommand, deleteSheet) => {
  // MY_COMMAND should takes DELETE_SHEET into account as DELETE_SHEET is arrived first
  if (myCommand.sheetId === deleteSheet.sheetId) {
    return undefined;
  }
  return myCommand;
});
```

### History

In addition to a transformation, an "inverse function" is required for commands which transforms other commands. The inverse function takes as input a `CoreCommand` and should return a list of `CoreCommands`. Those commands should have the inverse behavior as the original commands when used in transformation.

```txt
Given:
  A = a command
  B = another command
  A' = A transformed with B
  B⁻¹ = inverse of B
The inverse function should yield a command B⁻¹ such that:
  A === (A' transformed with B⁻¹)
```

In other words, transforming with `B` then transforming the result with `B⁻¹` should have no effect and yield the original command.

The inverse function is used during a selective undo to transform commands executed after the undo-ed command as if the command had never been executed.

Here is the way to declare it.

```js
const { inverseCommandRegistry } = o_spreadsheet.registries;

inverseCommandRegistry.add("CREATE_SHEET", (cmd) => {
  return [{ type: "DELETE_SHEET", sheetId: cmd.sheetId }];
});
```
