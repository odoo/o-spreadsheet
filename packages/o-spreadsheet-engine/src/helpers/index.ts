export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
