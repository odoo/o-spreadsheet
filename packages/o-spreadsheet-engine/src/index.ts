export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
