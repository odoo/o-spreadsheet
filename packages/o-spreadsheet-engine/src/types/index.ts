export * from "./base";
export * from "./errors";
export * from "./functions";
export * from "./history";
export * from "./isMatrix";
export * from "./locale";
export * from "./validator";
