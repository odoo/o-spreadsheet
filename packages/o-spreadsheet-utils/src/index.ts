export * from "./translation";
export * from "./uuid";
