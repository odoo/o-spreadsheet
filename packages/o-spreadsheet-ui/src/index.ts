import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}