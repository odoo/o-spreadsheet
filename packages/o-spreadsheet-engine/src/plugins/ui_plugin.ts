import { Session } from "../collaborative/session";
import { StateObserver } from "../state_observer";
import { ClientPosition } from "../types/collaborative/session";
import { Command, CommandDispatcher, CommandResult, CoreCommand } from "../types/commands";
import { Currency } from "../types/currency";
import { Getters } from "../types/getters";
import { HistoryChange } from "../types/history2";
import { Color } from "../types/misc";
import { ModelConfig } from "../types/model";
import { GridRenderingContext, LayerName } from "../types/rendering";
import { SelectionStreamProcessor } from "../types/selection_stream_processor";
import { ExcelWorkbookData } from "../types/workbook_data";
import { BasePlugin } from "./base_plugin";

export type UIActions = Pick<ModelConfig, "notifyUI" | "raiseBlockingErrorUI">;

export interface UIPluginConfig {
  readonly getters: Getters;
  readonly stateObserver: StateObserver<CoreCommand, HistoryChange>;
  readonly dispatch: CommandDispatcher["dispatch"];
  readonly canDispatch: CommandDispatcher["dispatch"];
  readonly selection: SelectionStreamProcessor;
  readonly moveClient: (position: ClientPosition) => void;
  readonly uiActions: UIActions;
  readonly custom: ModelConfig["custom"];
  readonly session: Session;
  readonly defaultCurrency?: Partial<Currency>;
  readonly customColors: Color[];
  readonly external: ModelConfig["external"];
}

export interface UIPluginConstructor {
  new (config: UIPluginConfig): UIPlugin;
  layers: Readonly<LayerName[]>;
  getters: readonly string[];
}

/**
 * UI plugins handle any transient data required to display a spreadsheet.
 * They can draw on the grid canvas.
 */
export class UIPlugin<State = any> extends BasePlugin<
  State,
  Command,
  CommandResult,
  HistoryChange,
  ExcelWorkbookData
> {
  static layers: Readonly<LayerName[]> = [];

  protected getters: Getters;
  protected ui: UIActions;
  protected selection: SelectionStreamProcessor;
  protected dispatch: CommandDispatcher["dispatch"];
  protected canDispatch: CommandDispatcher["dispatch"];

  constructor({
    getters,
    stateObserver,
    dispatch,
    canDispatch,
    uiActions,
    selection,
  }: UIPluginConfig) {
    super(stateObserver, CommandResult.Success);
    this.getters = getters;
    this.ui = uiActions;
    this.selection = selection;
    this.dispatch = dispatch;
    this.canDispatch = canDispatch;
  }

  // ---------------------------------------------------------------------------
  // Grid rendering
  // ---------------------------------------------------------------------------

  drawLayer(ctx: GridRenderingContext, layer: LayerName) {}
}
