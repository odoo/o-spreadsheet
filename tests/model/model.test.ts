import { CommandResult, CorePlugin } from "../../src";
import { toZone } from "../../src/helpers";
import { Model, ModelConfig } from "../../src/model";
import { corePluginRegistry, featurePluginRegistry } from "../../src/plugins/index";
import { UIPlugin } from "../../src/plugins/ui_plugin";
import { Command, CommandTypes, CoreCommand, DispatchResult, coreTypes } from "../../src/types";
import { setupCollaborativeEnv } from "../collaborative/collaborative_helpers";
import { copy, selectCell, setCellContent } from "../test_helpers/commands_helpers";
import {
  getCell,
  getCellContent,
  getCellText,
  getEvaluatedCell,
} from "../test_helpers/getters_helpers";
import { addTestPlugin } from "../test_helpers/helpers";

describe("Model", () => {
  test("core plugin can refuse command from UI plugin", () => {
    class MyCorePlugin extends CorePlugin {
      allowDispatch(cmd: CoreCommand) {
        if (cmd.type === "UPDATE_CELL") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
    }
    let result: DispatchResult | undefined = undefined;
    class MyUIPlugin extends UIPlugin {
      handle(cmd: Command) {
        if (cmd.type === "COPY") {
          result = this.dispatch("UPDATE_CELL", {
            col: 0,
            row: 0,
            sheetId: this.getters.getActiveSheetId(),
            content: "hello",
          });
        }
      }
    }
    addTestPlugin(featurePluginRegistry, MyUIPlugin);
    addTestPlugin(corePluginRegistry, MyCorePlugin);
    const model = new Model();
    copy(model, "A1");
    expect(result).toBeCancelledBecause(CommandResult.CancelledForUnknownReason);
  });

  test("core plugin cannot refuse command from core plugin", () => {
    let result: DispatchResult | undefined = undefined;
    class MyCorePlugin extends CorePlugin {
      allowDispatch(cmd: CoreCommand) {
        if (cmd.type === "UPDATE_CELL") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
      handle(cmd: CoreCommand) {
        if (cmd.type === "CREATE_SHEET") {
          result = this.dispatch("UPDATE_CELL", {
            col: 0,
            row: 0,
            sheetId: cmd.sheetId,
            content: "Hello",
          });
        }
      }
    }
    addTestPlugin(corePluginRegistry, MyCorePlugin);
    const model = new Model();
    model.dispatch("CREATE_SHEET", { sheetId: "42", position: 1 });
    expect(result).toBeSuccessfullyDispatched();
    expect(getCellText(model, "A1", "42")).toBe("Hello");
  });

  test("UI plugin cannot refuse command from UI plugin", () => {
    let result: DispatchResult | undefined = undefined;
    class MyUIPlugin extends UIPlugin {
      allowDispatch(cmd: Command) {
        if (cmd.type === "PASTE") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
      handle(cmd: Command) {
        if (cmd.type === "COPY") {
          result = this.dispatch("PASTE", {
            target: [toZone("A2")],
          });
        }
      }
    }
    addTestPlugin(featurePluginRegistry, MyUIPlugin);
    const model = new Model();
    setCellContent(model, "A1", "copy&paste me");
    copy(model, "A1");
    expect(result).toBeSuccessfullyDispatched();
    expect(getCellText(model, "A2")).toBe("copy&paste me");
  });

  test("UI plugins cannot refuse core commands", () => {
    class MyUIPlugin extends UIPlugin {
      allowDispatch(cmd: Command) {
        if (cmd.type === "UPDATE_CELL") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
    }
    addTestPlugin(featurePluginRegistry, MyUIPlugin);
    const model = new Model();

    setCellContent(model, "A1", "hello");
    expect(getCellContent(model, "A1")).toBe("hello");
  });

  test("Core plugins allowDispatch don't receive UI commands", () => {
    const receivedCommands: CommandTypes[] = [];
    class MyCorePlugin extends CorePlugin {
      allowDispatch(cmd: CoreCommand): CommandResult {
        receivedCommands.push(cmd.type);
        return CommandResult.Success;
      }
    }
    addTestPlugin(corePluginRegistry, MyCorePlugin);
    const model = new Model();
    model.dispatch("COPY");
    expect(receivedCommands).not.toContain("COPY");
  });

  test("Core plugins handle don't receive UI commands", () => {
    const receivedCommands: CommandTypes[] = [];
    class MyCorePlugin extends CorePlugin {
      handle(cmd: CoreCommand) {
        receivedCommands.push(cmd.type);
      }
    }
    addTestPlugin(corePluginRegistry, MyCorePlugin);
    const model = new Model();
    model.dispatch("COPY");
    expect(receivedCommands).not.toContain("COPY");
  });

  test("canDispatch method is exposed and works", () => {
    class MyCorePlugin extends CorePlugin {
      allowDispatch(cmd: CoreCommand) {
        if (cmd.type === "CREATE_SHEET") {
          return CommandResult.CancelledForUnknownReason;
        }
        return CommandResult.Success;
      }
    }
    addTestPlugin(corePluginRegistry, MyCorePlugin);
    const model = new Model();
    expect(model.canDispatch("CREATE_SHEET", { sheetId: "42", position: 1 })).toBeCancelledBecause(
      CommandResult.CancelledForUnknownReason
    );
    expect(model.dispatch("CREATE_SHEET", { sheetId: "42", position: 1 })).toBeCancelledBecause(
      CommandResult.CancelledForUnknownReason
    );

    const sheetId = model.getters.getActiveSheetId();
    expect(
      model.canDispatch("UPDATE_CELL", { sheetId, col: 0, row: 0, content: "hey" })
    ).toBeSuccessfullyDispatched();
    expect(
      model.dispatch("UPDATE_CELL", { sheetId, col: 0, row: 0, content: "hey" })
    ).toBeSuccessfullyDispatched();
    corePluginRegistry.remove("myCorePlugin");
  });

  test("Can open a model in readonly mode", () => {
    const model = new Model({}, { mode: "readonly" });
    expect(model.getters.isReadonly()).toBe(true);
  });

  test("Some commands are not dispatched in readonly mode", () => {
    const model = new Model({}, { mode: "readonly" });
    expect(setCellContent(model, "A1", "hello")).toBeCancelledBecause(CommandResult.Readonly);
  });

  test("Moving the selection is allowed in readonly mode", () => {
    const model = new Model({}, { mode: "readonly" });
    expect(selectCell(model, "A15")).toBeSuccessfullyDispatched();
  });

  test("Can add custom elements in the config of model", () => {
    const model = new Model({}, { custom: "42" } as unknown as ModelConfig);
    expect(model["config"]["custom"]).toBe("42");
  });

  test("type property in command payload is ignored", () => {
    const model = new Model();
    const payload = {
      col: 0,
      row: 0,
      sheetId: model.getters.getActiveSheetId(),
      content: "hello",
      type: "greeting",
    };
    model.dispatch("UPDATE_CELL", payload);
    expect(getCell(model, "A1")?.content).toBe("hello");
  });

  test("Cannot add an already existing core getters", () => {
    class MyCorePlugin1 extends CorePlugin {
      static getters = ["getSomething"];

      getSomething() {}
    }

    class MyCorePlugin2 extends CorePlugin {
      static getters = ["getSomething"];

      getSomething() {}
    }
    addTestPlugin(corePluginRegistry, MyCorePlugin1);
    addTestPlugin(corePluginRegistry, MyCorePlugin2);

    expect(() => new Model()).toThrowError(`Getter "getSomething" is already defined.`);
  });

  test("Cannot add an already existing getters", () => {
    class MyUIPlugin1 extends UIPlugin {
      static getters = ["getSomething"];

      getSomething() {}
    }

    class MyUIPlugin2 extends UIPlugin {
      static getters = ["getSomething"];

      getSomething() {}
    }
    addTestPlugin(featurePluginRegistry, MyUIPlugin1);
    addTestPlugin(featurePluginRegistry, MyUIPlugin2);

    expect(() => new Model()).toThrowError(`Getter "getSomething" is already defined.`);
  });

  test("Replayed commands are not send to UI plugins", () => {
    let numberCall = 0;
    //@ts-ignore
    coreTypes.add("MY_CMD_1");
    //@ts-ignore
    coreTypes.add("MY_CMD_2");
    class MyUIPlugin extends UIPlugin {
      handle(cmd: Command) {
        //@ts-ignore
        if (cmd.type === "MY_CMD_2") {
          if (this.getters.getClient().id === "bob") {
            numberCall++;
          }
        }
      }
    }
    addTestPlugin(featurePluginRegistry, MyUIPlugin);

    class MyCorePlugin extends CorePlugin {
      public readonly state: number = 0;
      handle(cmd: CoreCommand) {
        //@ts-ignore
        if (cmd.type === "MY_CMD_1") {
          this.history.update("state", 1);
          //@ts-ignore
          this.dispatch("MY_CMD_2");
        }
      }
    }
    addTestPlugin(corePluginRegistry, MyCorePlugin);

    const { alice, bob, network } = setupCollaborativeEnv();
    network.concurrent(() => {
      setCellContent(alice, "A1", "Hello");
      //@ts-ignore
      bob.dispatch("MY_CMD_1");
    });
    expect(numberCall).toEqual(1);
  });

  test("Initial commands are not sent to UI plugins", () => {
    class MyUIPlugin extends UIPlugin {
      handle(cmd: Command) {
        if (cmd.type === "UPDATE_CELL") {
          throw new Error("Should not be called");
        }
      }
    }
    addTestPlugin(featurePluginRegistry, MyUIPlugin);
    const data = {
      sheets: [{ id: "sheet1" }],
      revisionId: "initialRevision",
    };
    const model = new Model(data, {}, [
      {
        type: "REMOTE_REVISION",
        nextRevisionId: "1",
        serverRevisionId: "initialRevision",
        commands: [
          {
            type: "UPDATE_CELL",
            sheetId: "sheet1",
            col: 0,
            row: 0,
            content: "=1+5",
          },
        ],
        clientId: "1",
        version: 1,
      },
    ]);
    expect(getEvaluatedCell(model, "A1").value).toBe(6);
  });

  test("Core commands which dispatch UPDATE_CELL should trigger evaluation", () => {
    //@ts-ignore
    coreTypes.add("MY_CMD_1");
    class MyCorePlugin extends CorePlugin {
      handle(cmd: CoreCommand) {
        //@ts-ignore
        if (cmd.type === "MY_CMD_1") {
          this.dispatch("UPDATE_CELL", {
            //@ts-ignore
            sheetId: cmd.sheetId,
            col: 0,
            row: 0,
            content: "=5",
          });
        }
      }
    }
    addTestPlugin(corePluginRegistry, MyCorePlugin);

    const { alice, bob, charlie } = setupCollaborativeEnv();
    setCellContent(alice, "A1", "=3");
    expect([alice, bob, charlie]).toHaveSynchronizedValue(
      (user) => getCellContent(user, "A1"),
      "3"
    );
    //@ts-ignore
    alice.dispatch("MY_CMD_1", { sheetId: alice.getters.getActiveSheetId() });
    expect([alice, bob, charlie]).toHaveSynchronizedValue(
      (user) => getCellContent(user, "A1"),
      "5"
    );
  });

  test("export formula with unbound zone stays unbound", () => {
    const modelData = {
      sheets: [
        {
          cells: {
            A1: { content: "=SUM(A3:3)" },
            A2: { content: "=SUM(A3:A)" },
          },
        },
      ],
    };
    const model = new Model(modelData);
    expect(model.exportData()).toMatchObject({
      sheets: [
        {
          cells: {
            A1: { content: "=SUM(A3:3)" },
            A2: { content: "=SUM(A3:A)" },
          },
        },
      ],
    });
  });
});
