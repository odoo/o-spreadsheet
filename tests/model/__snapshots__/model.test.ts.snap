// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Model export formula with unbound zone stays unbound 1`] = `
{
  "borders": {},
  "formats": {},
  "pivotNextId": 1,
  "pivots": {},
  "revisionId": "START_REVISION",
  "settings": {
    "locale": {
      "code": "en_US",
      "dateFormat": "m/d/yyyy",
      "decimalSeparator": ".",
      "formulaArgSeparator": ",",
      "name": "English (US)",
      "thousandsSeparator": ",",
      "timeFormat": "hh:mm:ss a",
    },
  },
  "sheets": [
    {
      "areGridLinesVisible": true,
      "cells": {
        "A1": {
          "content": "=SUM(A3:3)",
          "format": undefined,
          "style": undefined,
        },
        "A2": {
          "content": "=SUM(A3:A)",
          "format": undefined,
          "style": undefined,
        },
      },
      "colNumber": 26,
      "cols": {},
      "conditionalFormats": [],
      "dataValidationRules": [],
      "figures": [],
      "headerGroups": {
        "COL": [],
        "ROW": [],
      },
      "id": "Sheet1",
      "isVisible": true,
      "merges": [],
      "name": "Sheet1",
      "rowNumber": 100,
      "rows": {},
      "tables": [],
    },
  ],
  "styles": {},
  "uniqueFigureIds": true,
  "version": 16,
}
`;
